ixwebsocket/IXWebSocketTransport.cpp

```diff
@@ -176,33 +202,51 @@ namespace ix
         _onCloseCallback = onCloseCallback;
     }
 
-    // Only consider send time points for that computation.
-    // The receive time points is taken into account in Socket::poll (second parameter).
-    bool WebSocketTransport::heartBeatPeriodExceeded()
+    // Only consider send PING time points for that computation.
+    bool WebSocketTransport::pingIntervalExceeded()
     {
-        std::lock_guard<std::mutex> lock(_lastSendTimePointMutex);
+        if (_pingIntervalSecs <= 0)
+            return false;
+
+        std::lock_guard<std::mutex> lock(_lastSendPingTimePointMutex);
         auto now = std::chrono::steady_clock::now();
-        return now - _lastSendTimePoint > std::chrono::seconds(_heartBeatPeriod);
+        return now - _lastSendPingTimePoint > std::chrono::seconds(_pingIntervalSecs);
+    }
+
+    bool WebSocketTransport::pingTimeoutExceeded()
+    {
+        if (_pingTimeoutSecs <= 0)
+            return false;
+
+        std::lock_guard<std::mutex> lock(_lastReceivePongTimePointMutex);
+        auto now = std::chrono::steady_clock::now();
+        return now - _lastReceivePongTimePoint > std::chrono::seconds(_pingTimeoutSecs);
     }
 
     void WebSocketTransport::poll()
     {
+        if (_readyState == OPEN)
+        {
+            // if (1) ping tiemout is enabled and (2) duration exceeded the maximum delay, then close the connection
```

Comment:
typo: tiemout -> timeout

Reply:
The new code tries to send pings more frequently, which is a behavior change, but is necessary if we want to make sure that ping are properly acknowledged.

Is it not possible to put that conditional block in the callback, under the 'pollResult == PollResultType::Timeout' case ? That would make the code more coherent. Would the test still pass with this ?

Or maybe we'd execute that block inconditionally (and forget that PollResultType::Timeout). But I'd like to see if we can add this code inside the callback.

Reply:
We need to do this in all cases so we can’t put that inside AND in the Timeout condition, because if we have always data to process the Timeout condition will never be true and there won’t be a Ping sent nor a ping timeout check. We could maybe put that code inside the poll, but before and without checking the Timeout. It should work the same, and we also have to be sure that calling close() method inside poll is not a problem. 
The test should still pass

Reply:
Thanks for the explanation. 

1. You convinced me that the timeout cannot be used, for the case when the connection is very busy and the timeout never happen. Let's forget about using it.
2. We call close from within the callback already so that should be alright to call close within the callback. -> Can you try to move that block inside the callback ? Adding a small comment, which is a hash of what you said in this review would be good.

Reply:
No problem, I think you answered from my original comment received in the email, I updated it since

Reply:
I think I still prefer to have this code run inside the callback. If I end up being wrong we can just make a micro-change and move it out of there back to where you have it now.

This function will be called over and over, so 'before' and 'after' state will become irrelevant fast.

Reply:
Ok then I will move to in the poll tomorrow